### PR TITLE
Improve escape sequence handling in private names

### LINKED ISSUE #50851: PrivateName does not permit escape characters
Discovered by @amcasey while running TS on test262.

```ts
export class IdentifierNameWithEscape {
    \u{78}: number;

    constructor() {
        this.\u{78} = 0;
    }

    doThing() {
        this.x = 42;
    }
}

export class PrivateIdentifierWithEscape {
    #\u{78}: number;

    constructor() {
        this.#\u{78} = 0;
    }

    doThing() {
        this.#x = 42;
    }
}
```

Currently `IdentifierNameWithEscape` parses and type-checks correctly; however, `PrivateIdentifierWithEscape` does not even parse.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
This PR makes two changes:

1. It fixes how the parser acquires the name of a *PrivateIdentifier* from the scanner, preferring the "cooked" `tokenValue` after processing escape sequences rather than just reading the plain source (the "token text"). Previously, two private names with the same effective/cooked names would not resolve to each other unless their source text was verbatim identical.
2. It fixes the scanner, which previously didn't expect a that a hash/pound (`#`) could be followed by a backlash (`\`).

There is still a slight problem around the fact that under ES3 and ES5 (our defaults), we don't recognize extended escape sequences at all. Ideally we would improve this by gracefully parsing and issuing an error that we can't emit them or something. But what that means is that I'm not *yet* certain as to whether the language service issue discovered within test262 (under formatjs) in #50835 is actually fixed by this. That might need to be a follow-up.

Fixes #50851.

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by amcasey** [comment]
> There is still a slight problem around the fact that under ES3 and ES5 (our defaults), we don't recognize extended escape sequences at all. Ideally we would improve this by gracefully parsing and issuing an error that we can't emit them or something.

I'm not sure I follow.  Those defaults are about our _output_, aren't they?  We always parse according to the latest spec supported by that version of TS?  Also, why would we be unable to emit?  Wouldn't we just emit the actual unicode character instead of the escape sequence?

> But what that means is that I'm not _yet_ certain as to whether the language service issue discovered within test262 (under formatjs) in #50835 is actually fixed by this. That might need to be a follow-up.

Fortunately, you can replay the repro locally with your locally-build compiler. :smile:

**Comment by DanielRosenwasser** [comment]
> Fortunately, you can replay the repro locally with your locally-build compiler.

I work on a codespace and I'm not certain how easy it is to `wget` the replay.

> Those defaults are about our _output_, aren't they?

Of course you're right, and this is a ridiculous problem.

> We always parse according to the latest spec supported by that version of TS? 

Unfortunately this is not the case - check out the differences in the new es5/es2015 `.errors.txt` files. 

Also

https://github.com/microsoft/TypeScript/blob/812ebcf6e3aebfa72b976a6ae4d65929759a2867/src/compiler/scanner.ts#L1489

> Wouldn't we just emit the actual unicode character instead of the escape sequence?

In theory we should - but we don't because I believe we update the list of valid identifiers depending on the language version.

File: src/compiler/scanner.ts [hunk 1]
```diff
@@ -2052,12 +2052,38 @@ namespace ts {
...
+                        }
+
+                        if (isIdentifierStart(charAfterHash, languageVersion)) {
+                            pos++;
+                            // We're relying on scanIdentifier's behavior and adjusting the token kind after the fact.
+                            // Notably absent from this block is the fact that calling a function named "scanIdentifier",
+                            // but identifiers don't include '#', and that function doesn't deal with it at all.
+                            // This works because 'scanIdentifier' tries to reuse source characters and builds up substrings;
+                            // however, it starts at the 'tokenPos' which includes the '#', and will "accidentally" prepend the '#' for us.
+                            scanIdentifier(charAfterHash, languageVersion);
```
Comment (amcasey) [inline src/compiler/scanner.ts:2083]: This change looks consistent with unicode escape handling elsewhere in the scanner, but I'm not sure I understand why `scanIdentifier` doesn't handle them.  Is it illegal to start an identifier with an escape sequence?  (Maybe that would introduce an ambiguity, but none jumps to mind.)

Comment (amcasey) [reply src/compiler/scanner.ts:2083]: Am I reading [this](https://262.ecma-international.org/13.0/#prod-IdentifierStart) correctly? It looks like any identifier _can_ start with a unicode escape.  If so, why wouldn't the fix be in `scanIdentifier`?

File: src/compiler/parser.ts [hunk 2]
```diff
@@ -2281,7 +2281,7 @@ namespace ts {
 
         function parsePrivateIdentifier(): PrivateIdentifier {
             const pos = getNodePos();
-            const node = factory.createPrivateIdentifier(internPrivateIdentifier(scanner.getTokenText()));
+            const node = factory.createPrivateIdentifier(internPrivateIdentifier(scanner.getTokenValue()));
```
Comment (amcasey) [inline src/compiler/parser.ts:2284]: 😲

**Comment by amcasey** [comment]
> I work on a codespace and I'm not certain how easy it is to `wget` the replay. That's why I asked the bot to pack it.

I'm pretty sure `npm i` works, even if `wget` doesn't.

> > We always parse according to the latest spec supported by that version of TS?
> 
> Unfortunately this is not the case - check out the differences in the new es5/es2015 `.errors.txt` files.

TIL

> > Wouldn't we just emit the actual unicode character instead of the escape sequence?
> 
> In theory we should - but we don't because I believe we update the list of valid identifiers depending on the language version.

Still not following.  In each version, either the character is allowed in identifiers or it's not - it doesn't actually matter whether it was originally authored as an escape?

**Comment by DanielRosenwasser** [comment]
> In each version, either the character is allowed in identifiers or it's not - it doesn't actually matter whether it was originally authored as an escape?

Exactly - we already track whether something is an escape, and the emitter can handle that differently. I think whether the source character *translates* to a valid Unicode identifier in the target version is separate from whether we can parse out an extended escape.

I would rather do that in a separate PR though.

File: src/compiler/scanner.ts (hunk 1 above)
Comment (amcasey) [reply src/compiler/scanner.ts:2083]: @DanielRosenwasser Last question ⬆️

Comment (DanielRosenwasser) [reply src/compiler/scanner.ts:2083]: `scanIdentifier` could handle them if we just checked if the first character was either an identifier start or a `\`, but right now the logic is to not advance if we can't get at least one complete identifier start.

Right now the handling for any of the following incomplete escape sequences...

```
\
\u
\u0
\u00
\u000
\u{}
```

is to not munch up these characters, and identify the `\` as an unknown token, followed by whatever.

As an extension, what the current code does with private fields is to make each of

```
#\
#\u
#\u0
#\u00
#\u000
#\u{}
```

an incomplete private field with the name `#`, followed by an unknown token `\` followed by whatever.

Arguably, private fields could diverge here for some better errors.

**Review by amcasey — APPROVED** [review]
Thanks!

**Pull request merged**